projects/components/src/download-file/service/file-download.service.ts

```diff
@@ -0,0 +1,130 @@
+import { Injectable, OnDestroy } from '@angular/core';
+import { Dictionary } from '@hypertrace/common';
+import { isEmpty, startCase } from 'lodash-es';
+import { BehaviorSubject, combineLatest, Observable, of, Subscription } from 'rxjs';
+import { catchError, map, take } from 'rxjs/operators';
+import { NotificationService } from '../../notification/notification.service';
+
+@Injectable()
+export class FileDownloadService implements OnDestroy {
+  private readonly downloadElement: HTMLAnchorElement;
+  private readonly subscriptions: Subscription = new Subscription();
+  private readonly fileDownloadEventSubject: BehaviorSubject<FileDownloadEventType | undefined> = new BehaviorSubject<
+    FileDownloadEventType | undefined
+  >(undefined);
+
+  public readonly fileDownloadEvent$: Observable<
+    FileDownloadEventType | undefined
+  > = this.fileDownloadEventSubject.asObservable();
+
+  public constructor(private readonly notificationService: NotificationService) {
+    this.downloadElement = this.createDownloadElement();
+  }
+
+  /**
+   * Downloads data as text content
+   * @param config Text download config
+   */
+  public downloadAsText(config: FileDownloadBaseConfig): void {
+    this.download({ ...config }, data => `data:text/plain;charset=utf-8,${encodeURIComponent(data)}`);
+  }
+
+  /**
+   * Downloads data as csv formatted content
+   * @param config Csv download config
+   */
+  public downloadAsCsv(config: CsvDownloadFileConfig): void {
+    // If given header is empty, then create header from the data keys
+    const header$ = isEmpty(config.header)
+      ? config.dataSource.pipe(
+          map(data => Object.keys(data[0] ?? [])),
+          map(keys => keys.map(startCase))
+        )
+      : of(config.header!);
+
+    // Value replacer for null and undefined values
+    const replacer = (_: string, value: string) => value ?? '-';
+
+    // Convert values into strings
+    const values$ = config.dataSource.pipe(
+      map(data => data.map(datum => Object.values(datum).map(value => JSON.stringify(value, replacer))))
+    );
+
+    // Create csv data as string
+    const csvData$ = combineLatest([header$, values$]).pipe(
+      map(([header, values]) => [header, ...values]), // Merge header and values
+      map(data => data.map(datum => datum.join(',')).join('\r\n')) // Join data to create a string
+    );
+
+    this.download({ ...config, dataSource: csvData$ }, csvData => {
+      const csvFile = new Blob([csvData], { type: 'text/csv' });
+
+      return window.URL.createObjectURL(csvFile);
+    });
+  }
+
+  public ngOnDestroy(): void {
+    this.downloadElement.remove();
```

Comment:
So, would it cancel the download if we moved out of the page?

Reply:
- If this happens before `downloadElement.click()`, then Yes as the flow is not completed. (Which I think is okay)
- After that(on browser download), I don't think so, I will check.